### ISSUE TITLE
Using annotation processor to improve startup and initial run

### ISSUE THREAD

**Issue by hc-codersatlas** [nielsbasjes/yauaa#135]
Hi, as the startup time (and first runs, which can be made faster via the test warmups) are quite long a good improvement would be to use an annotation processor to e.g. process / convert most of the YAML into a class that can be executed and save a lot of this time / overhead. Essentially moving the startup code to occur at compile time.

**Comment by nielsbasjes** [comment]
In the past I have done some benchmarks to find out where the bottle neck lies in the startup.
You can run for example the TestSerialization and TestSerializationFast test to get an indication of how much time the various steps take.

Roughly speaking:
The reading of the Yaml files takes about ~150ms. 
The parsing of all the rules in there takes about ~50 ms
The biggest part is used the recursive walking through the parsed rules and creating the (very large) datastructures that are needed to make the runtime really fast. (~2000ms)

Simply storing the entire datastructure and deserializing it would be loading about 50MB of data which takes about 570 ms. In addition this thing needs to be read from somewhere.

So the optimization you propose would (if I understand correctly) mean that out of the 2200ms the speedup would either be:
1. about 10% lower ( ~2000ms) when just making all the yaml data classes. 
2. about 70% lower ( ~ 600ms) when making the complete datastructure part of the jar file. 

The latter would however make the system less flexible AND make the distributed jar file about 50MB larger. Which I consider to be a bit too much. 

I'm really curious: In what usecase is this 2 second startup a problem that needs to be solved?

**Comment by hc-codersatlas** [comment]
Hi, as an annotation processor this wouldn't be part of the jar of the actual runtime. It'd be a separate jar that runs during only the compile / build phase to output this 50MB and only once on first run unless updated. Having a Gradle plugin would be to similar effect too. We run a lot of integration tests with this included in different Microservices and reducing the startup by 600ms would be a great saving.

**Comment by nielsbasjes** [comment]
I had a good thought about this.

The initialization phase (the one that takes 2 seconds) makes a different datastructure depending on which fields you are requesting. This is done because this way the runtime becomes faster if you request less fields. Also this allows adding additional custom rules yourself.
So just about every use case will yield a different datastructure.
For the general use case precomputing this during the build phase does not make sense.
Precomputing it for every possible usecase is impossible.

As you indicated because you are doing a large number of integration tests and having a speedup would be very useful for you. For this usecase I recommend you either 

- instantiate a static instance of the analyzer and reuse that over several tests. 
- create an instance, serialize it (look at this https://github.com/nielsbasjes/yauaa/blob/master/analyzer/src/test/java/nl/basjes/parse/useragent/TestSerialization.java ) and then for each test deserialize the already constructed instance from the byte array.

Looking at the pros and cons I choose not to implement this annotation processor idea.

**Comment by hc-codersatlas** [comment]
Hi, the serialization concept is not a bad way of approaching it! Having said that the built-in JDK serializer is slow. I tried to run this against [Kryo](https://github.com/EsotericSoftware/kryo) but as classes, such as Matcher are missing no-args constructor it won't compile. Is this something you can look to add or would it be out of scope as well? Thanks.

**Comment by nielsbasjes** [comment]
Yes, that is something that should be fixed! --> #138
